docs/design/topics.md

```diff
@@ -0,0 +1,168 @@
+# Topics
+
+Authors: [Philip Metzger](mailto:philipmetzger@bluewin.ch), [Noah Mayr](mailto:dev@noahmayr.com)
+ [Anton Bulakh](mailto:him@necaq.ua)
+
+## Summary
+
+Introduce Topics as a truly Jujutsu native way for topological branches, which 
+also replace the current bookmark concept for Git interop. As they have been
+documented to be confusing users coming from Git. They also supersede the 
+`[experimental-advance-branches]` config for those who currently use it, as 
+such a behavior will be built-in for Topics.
+
+Topics have been discussed heavily since their appearance in 
+[this Discussion][gh-discuss]. As Noah, Anton and I had a long 
+[Discord discussion][dc-thread] about them, which then also poured into the 
+[Topic issue][issue].
+
+## Prior work 
+
+Currently there only is Mercurial which has a implementation of 
+[Topics][hg-topics]. There also is the [Topic feature][gerrit-topics] in Gerrit,
+which groups commits with a single identifier.
+
+
+## Goals and non-goals
+
+### Goals
+
+The goals for this Project are small, see below.
+
+* Introduce the concept of native topological branches for Jujutsu.
+* Simplify Git interop by reducing the burden on Bookmarks.
+* Add Change metadata as a storage concept.
+* Remove the awkward `bookmark` to Git `branch` mapping.
+
+### Non-Goals
+
+* Making bookmarks unnecessary.
+
+## Overview
+
+Until now, Jujutsu had no native set of topological branches, just 
+[Bookmarks][bm] which interact poorly with Git's expectation of branches. 
+Topics on the otherhand are can be made to represent Git branches as users 
+expect them, see [Julia Evans poll][jvns-poll]. They also allow us to 
+seamlessly take over the [tracking-branches][tb] concept.
+
+Other use-cases they're useful for are representing a set of 
+[archived commits][archived] or even a [checkout history][checkout].
+
+
+#### Behavior
+
+A Topic is a set of Changes marked with a name, which infectiously moves 
+forward with each descendant you create. All Changes without a named topic are
+implicitly in a anonymous topic, which gets separately tracked as soon as you 
+send it in for review (so materialize it as a Git branch). A Topic may have
+multiple heads. A single Change may belong to multiple Topics, which means 
+that they also can get exported as overlapping Git branches. 
+
+
+TODO: Example here
+
+#### Major Use-Cases 
+
+
+##### Obsoletion of `[experimental.auto-advance-bookmarks]`
+
+Since Topics are infectious by nature, they can perfectly map to a Git branch.
```

Comment:
how would a topic with multiple heads map to a single Git branch (or maybe multiple branches)?

Reply:
This sentence is unclear to me as well. I'm guessing what it's supposed to mean is that topics "perfectly map to" what people _expect from_ Git branches, rather than that they'd actually be implemented as a 1:1 equivalent for Git branches. But since Git branches don't actually work the way Topics are described (i.e. they are not merely "infectious" but actually implicitly apply to all ancestor commits, and they cannot have multiple heads), they don't actually "map perfectly" to Git branches as understood by people who deeply understand the Git model (which is probably the minority of Git users, but certainly a nontrivial contingent).

Certainly this sentence _cannot_ mean that topics are always represented in the git storage backend as branches, since that would be incompatible with both of the differences cited above.

Reply:
> how would a topic with multiple heads map to a single Git branch (or maybe multiple branches)?

Each separate head will get its own Git branch. There's also the variant of connecting them, but I do not think that is desirable, as the different heads should represent different kind of logical work.  

> This sentence is unclear to me as well. I'm guessing what it's supposed to mean is that topics "perfectly map to" what people _expect from_ Git branches, rather than that they'd actually be implemented as a 1:1 equivalent for Git branches. But since Git branches don't actually work the way Topics are described (i.e. they are not merely "infectious" but actually implicitly apply to all ancestor commits, and they cannot have multiple heads), they don't actually "map perfectly" to Git branches as understood by people who deeply understand the Git model (which is probably the minority of Git users, but certainly a nontrivial contingent).

That is correct. But sharing the conceptual behavior never meant that we as Jujutsu VCS need to represent them the same internally.

Reply:
Pushing the multiple heads of a topic as `foo-1`, `foo-2`, etc., is surprising to me. I was expecting topics to work roughly like this:

- A commit can manually be labeled a part of topic "foo".
- A new commit created on top of that commit will automatically be labeled "foo".
- If `heads(topic(foo))` has a single revision, the topic may be pushed as a git branch named "foo" with a command like `jj git push -t foo`
- If `heads(topic(foo))` has more than one revision, pushing the topic fails with an error (similar to hg refusing to create multiple remote heads)

I'm not sure why I would ever want to push a topic and get branches `foo-1`, `foo-2`, etc. The only reason might be to create a remote backup of my work, but I don't expect topics to fix this problem (it's something to solve for any anonymous head).

I think forges like Gerrit would support pushing multiple heads in the same topic, but I would expect `jj gerrit` to handle this.

I also don't know how creating multiple branches for a topic works as the graph changes over time. Suppose I had commits B and C in a single topic, both descending from A, and I pushed, creating branches `foo-1` and `foo-2`. Then I create a merge commit D also in the topic. Does pushing delete `foo-1` and `foo-2`, and replace them with just `foo`?

```
D
|\
B |  foo-1@origin
| C  foo-2@origin
|/
A
```